.github/workflows/publish-python.yml

```diff
@@ -0,0 +1,153 @@
+name: Publish Python Bindings
+on:
+  workflow_call:
+    inputs:
+      ref:
+        description: 'commit/tag/branch reference'
+        required: true
+        type: string
+      package-version:
+        description: 'version for the python package (MAJOR.MINOR.BUILD)'
+        required: true
+        type: string
+      publish:
+        description: 'value indicating whether to publish to pypi.'
+        required: true
+        type: boolean
+        default: false
+    secrets:
+      PYPI_API_TOKEN:
+        description: 'api token to authenticate to pypi'
+        required: true
+
+jobs:
+  build-macos-wheels:
+    runs-on: macos-latest
+    strategy:
+      matrix:
+        python: ["3.8", "3.9", "3.10"]
+    steps:
+      - name: Checkout breez-sdk repo
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ inputs.ref || github.sha }}
+
+      - name: "Install Python"
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python }}
+
+      - uses: actions/download-artifact@v3
+        with:
+          name: sdk-bindings-darwin-universal
+          path: libs/sdk-bindings/bindings-python/src/breez_sdk
+
+      - uses: actions/download-artifact@v3
+        with:
+          name: bindings-python
+          path: libs/sdk-bindings/bindings-python/src/breez_sdk
+
+      - name: Clean up downloaded files
+        run: |
+          rm -f libs/sdk-bindings/bindings-python/src/breez_sdk/*.a
+          ls -R libs/sdk-bindings/bindings-python
+
+      - name: Update package version
+        if: ${{ inputs.package-version }}
+        working-directory: libs/sdk-bindings/bindings-python
+        run: sed -i.bak -e 's/    version=".*",/    version="${{ inputs.package-version }}",/' setup.py
+
+      - name: Install dependencies
+        working-directory: libs/sdk-bindings/bindings-python
+        run: pip3 install wheel setuptools
+
+      - name: Build wheel
+        working-directory: libs/sdk-bindings/bindings-python
+        run: python3 setup.py bdist_wheel --plat-name macosx_11_0_universal2 --verbose
+
+      - name: List wheel contents
+        working-directory: libs/sdk-bindings/bindings-python/dist
+        run: python3 -m zipfile --list *.whl || true
+
+      - name: Archive the wheel
+        uses: actions/upload-artifact@v3
+        with:
+          name: python-wheel-${{ matrix.python }}-macos
+          path: libs/sdk-bindings/bindings-python/dist/*.whl
+
+  build-linux-wheels:
+    runs-on: ubuntu-20.04
+    container:
+      image: quay.io/pypa/manylinux2014_x86_64
```

Comment:
Good question, I could've provided more context here.

Basically, to build a Python wheel that works on as many versions of Linux as possible, the OS you build it on needs to meet certain criteria. E.g. it should be rather old. I have to say I wouldn't call myself a Python expert so I followed mostly what the PyPa project (the people that build and maintain things like `pip` or the `wheel` distribution format) recommend and used one of their docker images to do the build. If you or anyone of the team have a recommendation for a different appraoch, let me know. Would be happy to adapt the workflow.

> Building manylinux-compatible wheels is not trivial; as a general rule, binaries built on one Linux distro will only work on other Linux distros that are the same age or newer. Therefore, if we want to make binaries that run on most Linux distros, we have to use an old enough distro.
Rather than forcing you to install an old distro yourself, install Python, etc., we provide [Docker](https://docker.com/) images where we've done the work for you. The images are uploaded to [quay.io](https://quay.io/organization/pypa) and are tagged for repeatable builds.

Source: https://github.com/pypa/manylinux#docker-images

One thing to note here is while PyPa seems to be a big, reputable project using their container might introduce some unknowns into the build process sine we don't know for sure if the image they publish is really what they say it is. This might theoretically have security implications.

Reply:
Thanks for the clear explanation!

One of the issues I see is that the sdk binaries are built on ubuntu 20.04 right now. They depend on the glibc version of the platform. So you may be able to install the python package on that machine, but the binaries won't work, because the glibc version is too low. So probably we are best off compiling the python package on ubuntu 20.04 as well, so at least every python package is able to run on that linux platform.

Reply:
This is a good point! 

I checked and it seems Ubuntu 20.04 ships with glibc version 2.31 (some other sources say 2.27 but the official `ubuntu:20.04` docker base image comes with 2.31). The [manylinux2014](https://peps.python.org/pep-0599/#the-manylinux2014-policy) tag dictates that shared libraries use glibc 2.17 or less. That means we're theoretically in violation here if we build the Linux bindings on Ubuntu 20.04. Since GitHub doesn't provide an older image changing the build environment is not an option.

To avoid limiting our Python package to work only on Ubuntu systems though, if I understand correctly we can tag the wheel as `manylinux_2_31`:

> This tag is a promise: the wheel’s creator promises that the wheel will work on any mainstream Linux distro that uses glibc version ${GLIBCMAJOR}.${GLIBCMINOR} or later
>
>[PEP 600](https://peps.python.org/pep-0600/#specification)

Looking at the dependencies of our shared library:

``` shell
root@bb8300119f49:/app# ldd libbreez_sdk_bindings.so
        linux-vdso.so.1 (0x0000ffff8ee51000)
        libstdc++.so.6 => /lib/aarch64-linux-gnu/libstdc++.so.6 (0x0000ffff8dde1000)
        libgcc_s.so.1 => /lib/aarch64-linux-gnu/libgcc_s.so.1 (0x0000ffff8ddbd000)
        libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000ffff8dd8c000)
        libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000ffff8dce1000)
        libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000ffff8dccd000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffff8db5a000)
        /lib/ld-linux-aarch64.so.1 (0x0000ffff8ee21000)
```

These are all fine according to the whitelist in the [auditwheel](https://github.com/pypa/auditwheel/blob/main/src/auditwheel/policy/manylinux-policy.json#L324-L384) tool. I also checked the symbol versions in our shared library using `readelf -s libbreez_sdk_bindings` and it looks like we meet all the criteria listed by [auditwheel](https://github.com/pypa/auditwheel/blob/main/src/auditwheel/policy/manylinux-policy.json#L324-L384).

Given that Ubuntu 20.04 uses glibc 2.31, we should be fine to build the wheel there as well and tag it as `manylinux_2_31`. I will check if that works and report back.

Reply:
Sounds good!

Reply:
Alright, I changed it so that wheels are now built on `ubuntu:20.04` and tagged as compliant with any Linux distro that has `glibc >= 2.31` which is what's used by `ubuntu:20.04` as far as I can tell. That way there shouldn't be any issues at runtime with incompliant glibc versions.

I used this test script:

``` python
import breez_sdk

seed = breez_sdk.mnemonic_to_seed(
    "<12 words>"
)

print(seed)
```

and ran it inside these containers:

- `amd64/ubuntu:latest`
- `arm64v8/ubuntu:latest`
- `amd64/fedora:35`
- `arm64v8/fedora:35`

Worked 4 times. 🎉  Of course that's not all possible combinations we could test here but at least it shows that it seems to work fine on somewhat recent Debian and RPM based distros.